python/cuspatial/cuspatial/core/binops/binops.py

```diff
@@ -0,0 +1,409 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+from abc import ABC, abstractmethod
+
+import cudf
+
+from cuspatial.core._column.geocolumn import GeoColumn
+from cuspatial.core.binops.contains import contains_properly
+from cuspatial.utils.column_utils import (
+    contains_only_linestrings,
+    contains_only_multipoints,
+    contains_only_points,
+    contains_only_polygons,
+    has_same_dimension,
+)
+
+
+class _binop(ABC):
+    @abstractmethod
+    def preprocess(self, op, lhs, rhs):
+        """Preprocess the input data for the binary operation. This method
+        should be implemented by subclasses. Preprocess and postprocess are
+        used to implement the discrete math rules of the binary operations.
+
+        Notes
+        -----
+        Should only be called once.
+
+        Parameters
+        ----------
+        op : str
+            The binary operation to perform.
+        lhs : GeoSeries
+            The left-hand-side of the GeoSeries-level binary operation.
+        rhs : GeoSeries
+            The right-hand-side of the GeoSeries-level binary operation.
+
+        Returns
+        -------
+        GeoSeries
+            The left-hand-side of the internal binary operation, may be
+            reordered.
+        GeoSeries
+            The right-hand-side of the internal binary operation, may be
+            reordered.
+        """
+        pass
+
+    @abstractmethod
+    def postprocess(self, op, point_indices, point_result):
+        """Postprocess the output data for the binary operation. This method
+        should be implemented by subclasses.
+
+        Postprocess converts the raw results of the binary operation into
+        the final result. This is where the discrete math rules are applied.
+
+        Parameters
+        ----------
+        op : str
+            The binary operation to post process. Determines for example the
+            set operation to use for computing the result.
+        processed : cudf.Series
+            The data after applying the fundamental binary operation.
+
+        Returns
+        -------
+        cudf.Series
+            The output of the post processing, True/False results for
+            the specified binary op.
+        """
+        pass
+
+    def __init__(self, op, lhs, rhs, align=True):
```

Comment:
It comes strange to me that you would perform the heavy computation inside the init function. Usually in init you would just store the necessary state for the object and do compute in `__call__`.

Reply:
That's a good point. I should do something that is more obvious instead of something weird, which is to compute all the results in the `__init__` and then just return them in `__call__`. There may have been a design decision there that I'll encounter when I refactor it.

Reply:
Ok, I moved the computation into `__call__` and cleaned up the factoring another iteration.